### PR TITLE
Correlation id package throws ArgumentException

### PR DESCRIPTION
**Pull request by maartenba**
After looking at the logs today, I found the correlation id provider was causing several ArgumentExceptions.

Reason is that we share and reuse the HttpClient, while the correlation library expects a new client per request. In shared mode, the library tries to set the header more than once (in a non-thread safe way), resulting in the above exception.

I just created an update that removes the library:
• A custom correlating handler was created that sets the correlation ID in the logical call context
o This also allows us to pass the correlation ID to Serilog when we get around to adding it
• Whenever an HttpClient uses the CorrelatingHttpClientHandler, the correlation ID will be used when available
• The default Web API correlaton id property can be used ()

Less ceremony (no more SetCorrelationId()) and should be thread safe now.

Please have a look at the (and run a few requests against it if you can)

/cc @xavierdecoster @skofman1 @joyhui

### CODE REVIEW

File: src/NuGet.Services.Search.Client/packages.config [hunk 1]
```diff
@@ -1,11 +1,10 @@
...
 <packages>
   <package id="Microsoft.AspNet.WebApi.Client" version="5.2.3" targetFramework="net452" />
-  <package id="Microsoft.AspNet.WebApi.Core" version="5.1.2" targetFramework="net452" />
+  <package id="Microsoft.AspNet.WebApi.Core" version="5.2.3" targetFramework="net452" />
   <package id="Microsoft.Bcl" version="1.1.10" targetFramework="net452" />
   <package id="Microsoft.Bcl.Build" version="1.0.21" targetFramework="net452" />
   <package id="Microsoft.Net.Http" version="2.2.29" targetFramework="net452" />
   <package id="Newtonsoft.Json" version="6.0.8" targetFramework="net452" />
   <package id="NuGet.Services.Platform.Client" version="3.0.29-r-master" targetFramework="net452" />
-  <package id="WebApiCorrelator" version="1.1.0.0" targetFramework="net452" />
```
Comment (xavierdecoster) [inline src/NuGet.Services.Search.Client/packages.config]: Less external dependencies #ftw :)

File: tests/NuGetGallery.Facts/packages.config [hunk 2]
```diff
@@ -5,7 +5,6 @@
   <package id="entityframework" version="6.1.3" targetFramework="net461" />
   <package id="Glimpse" version="1.8.6" targetFramework="net461" />
   <package id="Glimpse.AspNet" version="1.9.2" targetFramework="net461" />
-  <package id="jQuery" version="1.4.1" targetFramework="net461" />
```
Comment (xavierdecoster) [inline tests/NuGetGallery.Facts/packages.config]: Good riddance :)

**Comment by xavierdecoster** [comment]
Looks good! (also tested locally with simultaneous requests) :shipit:

File: tests/NuGetGallery.Facts/SearchClient/CorrelatingHttpClientHandlerFacts.cs [hunk 3]
```diff
@@ -0,0 +1,64 @@
...
+
+            // Assert
+            var request = inspectingHandler.Requests.FirstOrDefault();
+            Assert.NotNull(request);
+            Assert.True(request.Headers.Contains(WebApiCorrelationHandler.CorrelationIdHttpHeaderName));
+            Assert.Equal(correlationId.ToString(), request.Headers.GetValues(WebApiCorrelationHandler.CorrelationIdHttpHeaderName).FirstOrDefault());
+        }
+
+        [Fact]
+        public async Task DoesNotAddCorrelationIdToRequestWhenNotAvailable()
```
Comment (skofman1) [inline tests/NuGetGallery.Facts/SearchClient/CorrelatingHttpClientHandlerFacts.cs]: If the client doesn't provide a correlation id, don't we still want to generate one and propagate to search?

Comment (maartenba) [inline tests/NuGetGallery.Facts/SearchClient/CorrelatingHttpClientHandlerFacts.cs]: WebApiCorrelationHandler generetaes it. In the unit test I did want to make sure we don't regenerate here.

**Comment by skofman1** [comment]
In what scenarios do we reuse the HttpClient? I went through SearchClient, and ExternalSearchClient, and in that flow it's created per scope:
 builder.RegisterType<ExternalSearchService>()
                    .AsSelf()
                    .As<ISearchService>()
                    .As<IIndexingService>()
                    .InstancePerLifetimeScope();

**Comment by maartenba** [comment]
ExternalSearchService has a static SearchClient, which is where the reuse comes from.

**Comment by skofman1** [comment]
This code is really cool :)

**Comment by maartenba** [comment]
Thanks :-) Ship-worthy?

**Comment by skofman1** [comment]
:shipit:

**Pull request merged**